File: go/pkg/mechbase/sections.go
package mechbase

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
)

// Sections is the section registry resource.
type Sections struct {
	client *Client
	iid    int
}

// SectionInput is the create/upsert/update body for sections.
type SectionInput struct {
	Name       *string `json:"name,omitempty"`
	ExternalID *string `json:"external_id,omitempty"`
}

// ListSectionsOptions filters Sections.List. Zero values are omitted.
type ListSectionsOptions struct {
	Q      string
	Limit  int
	Offset int
}

// List returns a page of sections.
func (s *Sections) List(ctx context.Context, opts ListSectionsOptions) ([]Section, error) {
	q := url.Values{}
	if opts.Q != "" {
		q.Set("q", opts.Q)
	}
	addPaging(q, opts.Limit, opts.Offset)
	var env listEnvelope[Section]
	if err := s.client.doJSON(ctx, http.MethodGet, pathf(s.iid, "/sections"), q, nil, &env); err != nil {
		return nil, err
	}
	return env.Items, nil
}

// Get fetches one section by numeric id.
func (s *Sections) Get(ctx context.Context, sectionID int) (*Section, error) {
	var out Section
	if err := s.client.doJSON(ctx, http.MethodGet, pathf(s.iid, fmt.Sprintf("/sections/%d", sectionID)), nil, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Create creates a section.
func (s *Sections) Create(ctx context.Context, in SectionInput) (*Section, error) {
	var out Section
	if err := s.client.doJSON(ctx, http.MethodPost, pathf(s.iid, "/sections"), nil, in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Upsert creates or updates a section, matched by external_id.
func (s *Sections) Upsert(ctx context.Context, in SectionInput) (*Section, error) {
	var out Section
	if err := s.client.doJSON(ctx, http.MethodPut, pathf(s.iid, "/sections"), nil, in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Update partially updates a section by numeric id.
func (s *Sections) Update(ctx context.Context, sectionID int, in SectionInput) (*Section, error) {
	var out Section
	if err := s.client.doJSON(ctx, http.MethodPatch, pathf(s.iid, fmt.Sprintf("/sections/%d", sectionID)), nil, in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Delete removes a section; cascade also removes children.
func (s *Sections) Delete(ctx context.Context, sectionID int, cascade bool) (*DeleteResult, error) {
	q := url.Values{}
	if cascade {
		q.Set("cascade", "true")
	}
	var out DeleteResult
	if err := s.client.doJSON(ctx, http.MethodDelete, pathf(s.iid, fmt.Sprintf("/sections/%d", sectionID)), q, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: swift/Sources/Mechbase/Sections.swift
import Foundation

/// Create/upsert/update body for sections. Nil fields are omitted by the encoder.
public struct SectionInput: Encodable, Sendable {
    public var name: String?
    public var externalId: String?

    public init(name: String? = nil, externalId: String? = nil) {
        self.name = name
        self.externalId = externalId
    }

    enum CodingKeys: String, CodingKey {
        case name
        case externalId = "external_id"
    }
}

public struct Sections: Sendable {
    let scope: InstallationScope

    init(http: HTTPClient, installationId: Int) {
        self.scope = InstallationScope(http: http, installationId: installationId)
    }

    public func list(query: String? = nil, limit: Int = 50, offset: Int = 0) async throws -> [Section] {
        let q: [URLQueryItem] = [
            URLQueryItem(name: "q", value: query),
            URLQueryItem(name: "limit", value: String(limit)),
            URLQueryItem(name: "offset", value: String(offset)),
        ]
        return try await scope.http.get(scope.path("/sections"), query: q, as: Page<Section>.self).items
    }

    public func get(id: Int) async throws -> Section {
        try await scope.http.get(scope.path("/sections/\(id)"), as: Section.self)
    }

    public func create(_ input: SectionInput) async throws -> Section {
        try await scope.http.postJSON(scope.path("/sections"), body: input, as: Section.self)
    }

    public func upsert(_ input: SectionInput) async throws -> Section {
        try await scope.http.putJSON(scope.path("/sections"), body: input, as: Section.self)
    }

    public func update(id: Int, _ input: SectionInput) async throws -> Section {
        try await scope.http.patchJSON(scope.path("/sections/\(id)"), body: input, as: Section.self)
    }

    public func delete(id: Int, cascade: Bool = false) async throws -> DeleteResult {
        let q = cascade ? [URLQueryItem(name: "cascade", value: "true")] : []
        return try await scope.http.delete(scope.path("/sections/\(id)"), query: q, as: DeleteResult.self)
    }
}
